### DISCUSSION TITLE
Full path to plugin

### DISCUSSION THREAD

**Discussion by dshaw619** [gitbls/sdm#63] [Q&A]
The [sdm Plugins Wiki page](https://github.com/gitbls/sdm/wiki/sdm-Plugins#plugin-overview) says "You can also specify the plugin name with a full path. sdm will copy the plugin to /usr/local/sdm/local-plugin if it does not exist or the one specified is newer than the one in local_plugins."  This happens correctly if the plugin does not exist in /usr/local/sdm/local-plugins but if I make changes to the plugin in my directory where I have all my configuration scripts and files, it does not get copied and the customization runs with the older, incorrect plugin.  If I delete the plugin script from /usr/local/sdm/local-plugins and run the exact same --customize command, the newer plugin I specify gets used as desired.

```
sudo /usr/local/sdm/sdm --customize \
--wpa $HOME/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/wpa_supplicant.conf \
--keymap us \
--locale en_US.UTF-8 \
--timezone America/Los_Angeles \
--wifi-country US \
--disable piwiz \
--svc-disable userconfig \
--bootset serial=0,boot_behaviour=B4,boot_wait=0 \
--nouser \
--fstab $HOME/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/fstab_append.txt \
--extend --xmb 1024 \
--poptions noupdate,noupgrade \
--plugin $HOME/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/sdm-plugin.sh:\
"test1=value1|test3=value3" \
--plugin-debug \
--restart \
~/Downloads/bullseye-armhf-customized.img
```

**Comment by gitbls** [comment]
The copy plugin or not code was wrong. Fixed in next update. Coming soon.

**Comment by gitbls** [comment]
This is fixed in V7.7. Please test and verify. Thanks!

**Reply by dshaw619** [reply]
In installed v7.7 and an updated plugin specified in my --customize command was correctly copied and used in Phase 0.

Something seems broken in Phase 1 though.  Looks like something related to the mounted disk vs. container environments (?).  I don't think (famous last words) I've changed anything that might have caused this so mentioning here before digging into it further.  
```
...
* Phase 0 Completed
* Enter image '/home/doug/Downloads/bullseye-armhf-customized.img' for Phase 1
? Internal error: File '/mnt/sdm//sdm-cparse' not found in readparams
/usr/local/sdm/sdm-phase1: line 175: logtoboth: command not found
/usr/local/sdm/sdm-phase1: line 176: logfreespace: command not found
... [many more such error messages]
```

**Reply by dshaw619** [reply]
From the comment associated with the latest code it seems something has change re. loadparams.  I checked the latest version of the plugin template in case I needed to change something.  Looks like the phase 1 section of my plugin (below) still matches the template.

```
...
elif [ "$phase" == "1" ]
then
    #
    # Phase 1 (in nspawn)
    #
    loadparams
    logtoboth "* Plugin $pfx: Start Phase 1"
    plugin_getargs $pfx "$args" "|test1|test2|" "|test1|"
    #logfreespace "at start of Plugin $pfx Phase 1"
    #
    # INSERT your Plugin Phase 1 customization stuff here
    # In Phase 1 all references to directories in the image can be direct
    #
    #logfreespace "at end of $pfx Phase 1"
    logtoboth "* Plugin $pfx: Complete Phase 1"
else
...
```

**Comment by gitbls** [comment]
Interesting. Of course I'm not seeing that on my systems (I would have fixed it!). A couple of questions:

- How did you install V7.7?
- Can you `sdm --explore bullseye-armhf-customized.img` do `grep sdmdir /etc/sdm/cparams` and post the output here?

I'm sure I'll have more questions, but starting with the most obvious issues first.

**Reply by dshaw619** [reply]
I installed v7.7 over the existing files by running the command at https://github.com/gitbls/sdm#install-sdm

```
doug@dv6-Notebook-PC:~$ sudo sdm --explore ~/Downloads/bullseye-armhf-customized.img
[sudo] password for doug:             
* Mount IMG '/home/doug/Downloads/bullseye-armhf-customized.img'
mount: /dev/loop0p2 mounted on /mnt/sdm.
mount: /dev/loop0p1 mounted on /mnt/sdm/boot.
% Configure chroot to use this IMG on your ' ARM' computer
  * sdm will run more slowly due to qemu emulation
* Enter IMG '/home/doug/Downloads/bullseye-armhf-customized.img'
root@dv6-Notebook-PC:/# 
root@dv6-Notebook-PC:/# grep sdmdir /etc/sdm/cparams
sdmdir:"/usr/local/sdm"
root@dv6-Notebook-PC:/#
```

I'm running Linux Mint 21 on an old HP laptop - x86_64 bits.

**Comment by gitbls** [comment]
Huh. I will admit that I did all my testing on the Pi, so didn't test cross-platform, but it _should_ just work. Let's see if it does when I actually test that.

...and it failed. Investigating.

**Reply by dshaw619** [reply]
I would completely understand if at some point you decided to declare it only for running on RPi OS.

**Reply by gitbls** [reply]
Nope, smells like a bug. 🤔

**Reply by gitbls** [reply]
Yep,it's obviously broken. Testing, then will post.

**Reply by gitbls** [reply]
OK...if you could just pull down sdm-cparse and drop it into the sdm directory (/usr/local/sdm) on your Mint system, it should be MUCH better now. LMK how it goes.

**Reply by dshaw619** [reply]
I copied the code from https://github.com/gitbls/sdm/blob/master/sdm-cparse, pasted into VS Code, and saved to /usr/local/sdm/sdm-cparse (I thought there was a button or pull-down menu item somewhere on Github to download or copy the full piece of code, but didn't seem to find it;  one discovers just how much code there is in cparse when scrolling to copy (:-) ).

Worked great!  Thanks.

**Comment by gitbls** [comment]
Excellent. Ia gree that it's not obvious how to copy a single file. I usually copy the zip file (Green 'code' button, then Download Zip) when I need something quickly.

BTW, the fact that sdm works cross-platform is a pretty cool feature (and does it seamlessly), so no plan to undo that. I just need to remember to test it 🙄 .
